File: agenkit/checkpointing/storage.py
"""
Checkpoint storage implementations.

Provides in-memory and file-based storage for checkpoints.
"""

from collections import defaultdict
from pathlib import Path

from .checkpoint import Checkpoint, CheckpointStorage


class InMemoryCheckpointStorage(CheckpointStorage):
    """
    In-memory checkpoint storage.

    Good for:
    - Testing
    - Development
    - Short-lived sessions

    Not suitable for:
    - Production (no persistence)
    - Long-running agents (lost on restart)
    """

    def __init__(self):
        # checkpoint_id -> Checkpoint
        self._checkpoints: dict[str, Checkpoint] = {}
        # session_id -> list of checkpoint_ids (ordered by timestamp)
        self._session_checkpoints: dict[str, list[str]] = defaultdict(list)

    async def save(self, checkpoint: Checkpoint) -> None:
        """Save checkpoint to memory."""
        self._checkpoints[checkpoint.checkpoint_id] = checkpoint

        # Add to session index
        if checkpoint.checkpoint_id not in self._session_checkpoints[checkpoint.session_id]:
            self._session_checkpoints[checkpoint.session_id].append(checkpoint.checkpoint_id)

            # Sort by timestamp (most recent first)
            self._session_checkpoints[checkpoint.session_id].sort(
                key=lambda cid: self._checkpoints[cid].timestamp, reverse=True
            )

    async def load(self, checkpoint_id: str) -> Checkpoint | None:
        """Load checkpoint from memory."""
        return self._checkpoints.get(checkpoint_id)

    async def list_checkpoints(self, session_id: str, limit: int | None = None) -> list[Checkpoint]:
        """List checkpoints for session."""
        checkpoint_ids = self._session_checkpoints.get(session_id, [])

        if limit:
            checkpoint_ids = checkpoint_ids[:limit]

        return [self._checkpoints[cid] for cid in checkpoint_ids]

    async def get_latest(self, session_id: str) -> Checkpoint | None:
        """Get latest checkpoint for session."""
        checkpoints = await self.list_checkpoints(session_id, limit=1)
        return checkpoints[0] if checkpoints else None

    async def delete(self, checkpoint_id: str) -> bool:
        """Delete checkpoint."""
        if checkpoint_id not in self._checkpoints:
            return False

        checkpoint = self._checkpoints[checkpoint_id]
        del self._checkpoints[checkpoint_id]

        # Remove from session index
        if checkpoint_id in self._session_checkpoints[checkpoint.session_id]:
            self._session_checkpoints[checkpoint.session_id].remove(checkpoint_id)

        return True

    async def delete_session(self, session_id: str) -> int:
        """Delete all checkpoints for session."""
        checkpoint_ids = self._session_checkpoints.get(session_id, [])
        count = len(checkpoint_ids)

        for checkpoint_id in checkpoint_ids:
            del self._checkpoints[checkpoint_id]

        del self._session_checkpoints[session_id]

        return count

    async def get_checkpoint_history(
        self, checkpoint_id: str, max_depth: int = 10
    ) -> list[Checkpoint]:
        """Get checkpoint history by following parent links."""
        history = []
        current_id = checkpoint_id

        for _ in range(max_depth):
            checkpoint = await self.load(current_id)
            if not checkpoint:
                break

            history.append(checkpoint)

            if not checkpoint.parent_checkpoint_id:
                break

            current_id = checkpoint.parent_checkpoint_id

        return history

    def get_stats(self) -> dict:
        """Get storage statistics."""
        return {
            "total_checkpoints": len(self._checkpoints),
            "total_sessions": len(self._session_checkpoints),
            "checkpoints_per_session": {
                session_id: len(checkpoint_ids)
                for session_id, checkpoint_ids in self._session_checkpoints.items()
            },
        }


class FileCheckpointStorage(CheckpointStorage):
    """
    File-based checkpoint storage.

    Stores each checkpoint as a JSON file on disk for persistence.

    Directory structure:
        checkpoint_dir/
            {session_id}/
                {checkpoint_id}.json
                {checkpoint_id}.json
                ...

    Good for:
    - Production (persistent)
    - Single-machine deployments
    - Development with persistence

    Example:
        >>> storage = FileCheckpointStorage("./checkpoints")
        >>> await storage.save(checkpoint)
    """

    def __init__(self, checkpoint_dir: str = "./checkpoints"):
        """
        Initialize file-based storage.

        Args:
            checkpoint_dir: Directory to store checkpoints
        """
        self.checkpoint_dir = Path(checkpoint_dir)
        self.checkpoint_dir.mkdir(parents=True, exist_ok=True)

    def _get_session_dir(self, session_id: str) -> Path:
        """Get directory for session checkpoints."""
        session_dir = self.checkpoint_dir / session_id
        session_dir.mkdir(exist_ok=True)
        return session_dir

    def _get_checkpoint_path(self, session_id: str, checkpoint_id: str) -> Path:
        """Get file path for checkpoint."""
        return self._get_session_dir(session_id) / f"{checkpoint_id}.json"

    async def save(self, checkpoint: Checkpoint) -> None:
        """Save checkpoint to file."""
        checkpoint_path = self._get_checkpoint_path(checkpoint.session_id, checkpoint.checkpoint_id)

        with open(checkpoint_path, "w") as f:
            f.write(checkpoint.to_json())

    async def load(self, checkpoint_id: str) -> Checkpoint | None:
        """Load checkpoint from file."""
        # Need to search through session directories
        for session_dir in self.checkpoint_dir.iterdir():
            if not session_dir.is_dir():
                continue

            checkpoint_path = session_dir / f"{checkpoint_id}.json"
            if checkpoint_path.exists():
                with open(checkpoint_path) as f:
                    return Checkpoint.from_json(f.read())

        return None

    async def list_checkpoints(self, session_id: str, limit: int | None = None) -> list[Checkpoint]:
        """List checkpoints for session."""
        session_dir = self._get_session_dir(session_id)

        if not session_dir.exists():
            return []

        # Load all checkpoints
        checkpoints = []
        for checkpoint_file in session_dir.glob("*.json"):
            with open(checkpoint_file) as f:
                checkpoint = Checkpoint.from_json(f.read())
                checkpoints.append(checkpoint)

        # Sort by timestamp (most recent first)
        checkpoints.sort(key=lambda c: c.timestamp, reverse=True)

        if limit:
            checkpoints = checkpoints[:limit]

        return checkpoints

    async def get_latest(self, session_id: str) -> Checkpoint | None:
        """Get latest checkpoint for session."""
        checkpoints = await self.list_checkpoints(session_id, limit=1)
        return checkpoints[0] if checkpoints else None

    async def delete(self, checkpoint_id: str) -> bool:
        """Delete checkpoint file."""
        # Search through session directories
        for session_dir in self.checkpoint_dir.iterdir():
            if not session_dir.is_dir():
                continue

            checkpoint_path = session_dir / f"{checkpoint_id}.json"
            if checkpoint_path.exists():
                checkpoint_path.unlink()
                return True

        return False

    async def delete_session(self, session_id: str) -> int:
        """Delete all checkpoints for session."""
        session_dir = self._get_session_dir(session_id)

        if not session_dir.exists():
            return 0

        # Count and delete checkpoint files
        checkpoint_files = list(session_dir.glob("*.json"))
        count = len(checkpoint_files)

        for checkpoint_file in checkpoint_files:
            checkpoint_file.unlink()

        # Remove session directory if empty
        try:
            session_dir.rmdir()
        except OSError:
            pass  # Directory not empty (might have other files)

        return count

    async def get_checkpoint_history(
        self, checkpoint_id: str, max_depth: int = 10
    ) -> list[Checkpoint]:
        """Get checkpoint history by following parent links."""
        history = []
        current_id = checkpoint_id

        for _ in range(max_depth):
            checkpoint = await self.load(current_id)
            if not checkpoint:
                break

            history.append(checkpoint)

            if not checkpoint.parent_checkpoint_id:
                break

            current_id = checkpoint.parent_checkpoint_id

        return history

    def get_stats(self) -> dict:
        """Get storage statistics."""
        stats = {
            "total_sessions": 0,
            "total_checkpoints": 0,
            "checkpoint_dir": str(self.checkpoint_dir),
            "disk_usage_bytes": 0,
        }

        for session_dir in self.checkpoint_dir.iterdir():
            if not session_dir.is_dir():
                continue

            stats["total_sessions"] += 1

            for checkpoint_file in session_dir.glob("*.json"):
                stats["total_checkpoints"] += 1
                stats["disk_usage_bytes"] += checkpoint_file.stat().st_size

        return stats

File: agenkit-ts/src/checkpointing/storage.ts
/**
 * Checkpoint storage implementations.
 *
 * Provides in-memory and file-based storage for checkpoints.
 */

import * as fs from 'fs';
import * as path from 'path';
import {
  Checkpoint,
  CheckpointStorage,
  checkpointFromJson,
  checkpointToJson,
} from './checkpoint';

/**
 * In-memory checkpoint storage.
 *
 * Good for:
 * - Testing
 * - Development
 * - Short-lived sessions
 *
 * Not suitable for:
 * - Production (no persistence)
 * - Long-running agents (lost on restart)
 *
 * Example:
 *   const storage = new InMemoryCheckpointStorage();
 *   await storage.save(checkpoint);
 *   const loaded = await storage.load(checkpoint.checkpointId);
 */
export class InMemoryCheckpointStorage implements CheckpointStorage {
  // checkpoint_id -> Checkpoint
  private checkpoints: Map<string, Checkpoint> = new Map();

  // session_id -> list of checkpoint_ids (ordered by timestamp)
  private sessionCheckpoints: Map<string, string[]> = new Map();

  /**
   * Save checkpoint to memory.
   */
  async save(checkpoint: Checkpoint): Promise<void> {
    this.checkpoints.set(checkpoint.checkpointId, checkpoint);

    // Add to session index
    if (!this.sessionCheckpoints.has(checkpoint.sessionId)) {
      this.sessionCheckpoints.set(checkpoint.sessionId, []);
    }

    const sessionCpts = this.sessionCheckpoints.get(checkpoint.sessionId)!;
    if (!sessionCpts.includes(checkpoint.checkpointId)) {
      sessionCpts.push(checkpoint.checkpointId);

      // Sort by timestamp (most recent first)
      sessionCpts.sort((a, b) => {
        const cptA = this.checkpoints.get(a)!;
        const cptB = this.checkpoints.get(b)!;
        return cptB.timestamp.getTime() - cptA.timestamp.getTime();
      });
    }
  }

  /**
   * Load checkpoint from memory.
   */
  async load(checkpointId: string): Promise<Checkpoint | undefined> {
    return this.checkpoints.get(checkpointId);
  }

  /**
   * List checkpoints for session.
   */
  async listCheckpoints(sessionId: string, limit?: number): Promise<Checkpoint[]> {
    const checkpointIds = this.sessionCheckpoints.get(sessionId) || [];

    const idsToLoad = limit ? checkpointIds.slice(0, limit) : checkpointIds;

    return idsToLoad
      .map((id) => this.checkpoints.get(id))
      .filter((cpt): cpt is Checkpoint => cpt !== undefined);
  }

  /**
   * Get latest checkpoint for session.
   */
  async getLatest(sessionId: string): Promise<Checkpoint | undefined> {
    const checkpoints = await this.listCheckpoints(sessionId, 1);
    return checkpoints[0];
  }

  /**
   * Delete checkpoint.
   */
  async delete(checkpointId: string): Promise<boolean> {
    const checkpoint = this.checkpoints.get(checkpointId);
    if (!checkpoint) {
      return false;
    }

    this.checkpoints.delete(checkpointId);

    // Remove from session index
    const sessionCpts = this.sessionCheckpoints.get(checkpoint.sessionId);
    if (sessionCpts) {
      const index = sessionCpts.indexOf(checkpointId);
      if (index !== -1) {
        sessionCpts.splice(index, 1);
      }
    }

    return true;
  }

  /**
   * Delete all checkpoints for session.
   */
  async deleteSession(sessionId: string): Promise<number> {
    const checkpointIds = this.sessionCheckpoints.get(sessionId) || [];
    const count = checkpointIds.length;

    for (const checkpointId of checkpointIds) {
      this.checkpoints.delete(checkpointId);
    }

    this.sessionCheckpoints.delete(sessionId);

    return count;
  }

  /**
   * Get checkpoint history by following parent links.
   */
  async getCheckpointHistory(
    checkpointId: string,
    maxDepth: number = 10,
  ): Promise<Checkpoint[]> {
    const history: Checkpoint[] = [];
    let currentId: string | undefined = checkpointId;

    for (let i = 0; i < maxDepth; i++) {
      if (!currentId) break;

      const checkpoint = await this.load(currentId);
      if (!checkpoint) break;

      history.push(checkpoint);

      if (!checkpoint.parentCheckpointId) break;

      currentId = checkpoint.parentCheckpointId;
    }

    return history;
  }

  /**
   * Get storage statistics.
   */
  getStats(): {
    totalCheckpoints: number;
    totalSessions: number;
    checkpointsPerSession: Record<string, number>;
  } {
    const checkpointsPerSession: Record<string, number> = {};

    for (const [sessionId, checkpointIds] of this.sessionCheckpoints.entries()) {
      checkpointsPerSession[sessionId] = checkpointIds.length;
    }

    return {
      totalCheckpoints: this.checkpoints.size,
      totalSessions: this.sessionCheckpoints.size,
      checkpointsPerSession,
    };
  }
}

/**
 * File-based checkpoint storage.
 *
 * Stores each checkpoint as a JSON file on disk for persistence.
 *
 * Directory structure:
 *   checkpoint_dir/
 *     {session_id}/
 *       {checkpoint_id}.json
 *       {checkpoint_id}.json
 *       ...
 *
 * Good for:
 * - Production (persistent)
 * - Single-machine deployments
 * - Development with persistence
 *
 * Example:
 *   const storage = new FileCheckpointStorage('./checkpoints');
 *   await storage.save(checkpoint);
 *   const loaded = await storage.load(checkpoint.checkpointId);
 */
export class FileCheckpointStorage implements CheckpointStorage {
  private checkpointDir: string;

  constructor(checkpointDir: string = './checkpoints') {
    this.checkpointDir = checkpointDir;

    // Create checkpoint directory if it doesn't exist
    if (!fs.existsSync(this.checkpointDir)) {
      fs.mkdirSync(this.checkpointDir, { recursive: true });
    }
  }

  /**
   * Get directory for session checkpoints.
   */
  private getSessionDir(sessionId: string): string {
    const sessionDir = path.join(this.checkpointDir, sessionId);
    if (!fs.existsSync(sessionDir)) {
      fs.mkdirSync(sessionDir, { recursive: true });
    }
    return sessionDir;
  }

  /**
   * Get file path for checkpoint.
   */
  private getCheckpointPath(sessionId: string, checkpointId: string): string {
    return path.join(this.getSessionDir(sessionId), `${checkpointId}.json`);
  }

  /**
   * Save checkpoint to file.
   */
  async save(checkpoint: Checkpoint): Promise<void> {
    const checkpointPath = this.getCheckpointPath(checkpoint.sessionId, checkpoint.checkpointId);
    const jsonStr = checkpointToJson(checkpoint);
    fs.writeFileSync(checkpointPath, jsonStr, 'utf-8');
  }

  /**
   * Load checkpoint from file.
   */
  async load(checkpointId: string): Promise<Checkpoint | undefined> {
    // Need to search through session directories
    if (!fs.existsSync(this.checkpointDir)) {
      return undefined;
    }

    const sessionDirs = fs.readdirSync(this.checkpointDir);

    for (const sessionDir of sessionDirs) {
      const sessionPath = path.join(this.checkpointDir, sessionDir);
      if (!fs.statSync(sessionPath).isDirectory()) {
        continue;
      }

      const checkpointPath = path.join(sessionPath, `${checkpointId}.json`);
      if (fs.existsSync(checkpointPath)) {
        const jsonStr = fs.readFileSync(checkpointPath, 'utf-8');
        return checkpointFromJson(jsonStr);
      }
    }

    return undefined;
  }

  /**
   * List checkpoints for session.
   */
  async listCheckpoints(sessionId: string, limit?: number): Promise<Checkpoint[]> {
    const sessionDir = this.getSessionDir(sessionId);

    if (!fs.existsSync(sessionDir)) {
      return [];
    }

    // Load all checkpoints
    const checkpoints: Checkpoint[] = [];
    const files = fs.readdirSync(sessionDir);

    for (const file of files) {
      if (!file.endsWith('.json')) continue;

      const checkpointPath = path.join(sessionDir, file);
      const jsonStr = fs.readFileSync(checkpointPath, 'utf-8');
      const checkpoint = checkpointFromJson(jsonStr);
      checkpoints.push(checkpoint);
    }

    // Sort by timestamp (most recent first)
    checkpoints.sort((a, b) => b.timestamp.getTime() - a.timestamp.getTime());

    if (limit) {
      return checkpoints.slice(0, limit);
    }

    return checkpoints;
  }

  /**
   * Get latest checkpoint for session.
   */
  async getLatest(sessionId: string): Promise<Checkpoint | undefined> {
    const checkpoints = await this.listCheckpoints(sessionId, 1);
    return checkpoints[0];
  }

  /**
   * Delete checkpoint file.
   */
  async delete(checkpointId: string): Promise<boolean> {
    // Search through session directories
    if (!fs.existsSync(this.checkpointDir)) {
      return false;
    }

    const sessionDirs = fs.readdirSync(this.checkpointDir);

    for (const sessionDir of sessionDirs) {
      const sessionPath = path.join(this.checkpointDir, sessionDir);
      if (!fs.statSync(sessionPath).isDirectory()) {
        continue;
      }

      const checkpointPath = path.join(sessionPath, `${checkpointId}.json`);
      if (fs.existsSync(checkpointPath)) {
        fs.unlinkSync(checkpointPath);
        return true;
      }
    }

    return false;
  }

  /**
   * Delete all checkpoints for session.
   */
  async deleteSession(sessionId: string): Promise<number> {
    const sessionDir = this.getSessionDir(sessionId);

    if (!fs.existsSync(sessionDir)) {
      return 0;
    }

    // Count and delete checkpoint files
    const files = fs.readdirSync(sessionDir).filter((f) => f.endsWith('.json'));
    const count = files.length;

    for (const file of files) {
      const filePath = path.join(sessionDir, file);
      fs.unlinkSync(filePath);
    }

    // Remove session directory if empty
    try {
      fs.rmdirSync(sessionDir);
    } catch {
      // Directory not empty (might have other files)
    }

    return count;
  }

  /**
   * Get checkpoint history by following parent links.
   */
  async getCheckpointHistory(
    checkpointId: string,
    maxDepth: number = 10,
  ): Promise<Checkpoint[]> {
    const history: Checkpoint[] = [];
    let currentId: string | undefined = checkpointId;

    for (let i = 0; i < maxDepth; i++) {
      if (!currentId) break;

      const checkpoint = await this.load(currentId);
      if (!checkpoint) break;

      history.push(checkpoint);

      if (!checkpoint.parentCheckpointId) break;

      currentId = checkpoint.parentCheckpointId;
    }

    return history;
  }

  /**
   * Get storage statistics.
   */
  getStats(): {
    totalSessions: number;
    totalCheckpoints: number;
    checkpointDir: string;
    diskUsageBytes: number;
  } {
    const stats = {
      totalSessions: 0,
      totalCheckpoints: 0,
      checkpointDir: this.checkpointDir,
      diskUsageBytes: 0,
    };

    if (!fs.existsSync(this.checkpointDir)) {
      return stats;
    }

    const sessionDirs = fs.readdirSync(this.checkpointDir);

    for (const sessionDir of sessionDirs) {
      const sessionPath = path.join(this.checkpointDir, sessionDir);
      if (!fs.statSync(sessionPath).isDirectory()) {
        continue;
      }

      stats.totalSessions++;

      const files = fs.readdirSync(sessionPath);
      for (const file of files) {
        if (!file.endsWith('.json')) continue;

        stats.totalCheckpoints++;
        const filePath = path.join(sessionPath, file);
        stats.diskUsageBytes += fs.statSync(filePath).size;
      }
    }

    return stats;
  }
}
